core/src/main/java/feign/Feign.java

```diff
@@ -219,4 +228,27 @@ public Feign build() {
       return new ReflectiveFeign(handlersByName, invocationHandlerFactory);
     }
   }
+
+  static class ResponseMappingDecoder implements Decoder {
+
+    private final ResponseMapper mapper;
+    private final Decoder delegate;
+
+    ResponseMappingDecoder(ResponseMapper mapper, Decoder decoder) {
+      this.mapper = mapper;
+      this.delegate = decoder;
+    }
+
+    @Override
+    public Object decode(Response response, Type type) throws IOException {
+      return delegate.decode(mapper.map(response, type), type);
+    }
+  }
+
+  /**
+   * Map function to apply to the response before decoding it.
+   */
+  public interface ResponseMapper {
```

Comment:
make this part a top-level type.

Also, please add an example in the javadoc.

ex. 
```
<pre>{@code
// your compelling example here:
}</pre>
```